qiita_db/processing_job.py

```diff
@@ -251,50 +251,58 @@ def launch_torque(env_script, start_script, url, job_id, job_dir,
-        # ends, most likely. Torque doesn't keep job state around forever, and
-        # creating a dependency on a job already completed has not been tested.
-        qsub_cmd.append("qsub -W depend=afterok:%s" % dependent_job_id)
-    else:
-        qsub_cmd.append("qsub")
-
-    qsub_cmd.append("%s %s" % (resource_params, fp))
-    qsub_cmd.append("-o %s/qsub-output.txt " % job_dir)
-    qsub_cmd.append("-e %s/qsub-error.txt" % job_dir)
-    # TODO: revisit epilogue
-    qsub_cmd.append("-l epilogue=/home/qiita/qiita-epilogue.sh")
-
-    qsub_cmd = ' '.join(qsub_cmd)
-
-    # stdX parameters added to support returning the Torque ID from qsub
-    # Popen() may also need universal_newlines=True
-    # may also need stdout = stdout.decode("utf-8").rstrip()
-    proc = Popen(qsub_cmd, shell=True, stdout=PIPE, stderr=PIPE)
+    with open(fp, 'w') as torque_job_file:
+        torque_job_file.write("\n".join(lines))
+        torque_job_file.close()
+
+        qsub_cmd = ['qsub']
+
+        if dependent_job_id:
+            # note that a dependent job should be submitted before the
+            # 'parent' job ends, most likely. Torque doesn't keep job state
+            # around forever, and creating a dependency on a job already
+            # completed has not been tested.
+            qsub_cmd.append("-W")
+            qsub_cmd.append("depend=afterok:%s" % dependent_job_id)
+
+        qsub_cmd.append("%s" % resource_params)
+        qsub_cmd.append("%s" % fp)
+        qsub_cmd.append("-o")
+        qsub_cmd.append("%s/qsub-output.txt" % job_dir)
+        qsub_cmd.append("-e")
+        qsub_cmd.append("%s/qsub-error.txt" % job_dir)
+        # TODO: revisit epilogue
+        qsub_cmd.append("-l")
+        qsub_cmd.append("epilogue=/home/qiita/qiita-epilogue.sh")
+
+        # stdX parameters added to support returning the Torque ID from qsub
+        # Popen() may also need universal_newlines=True
+        # may also need stdout = stdout.decode("utf-8").rstrip()
+        proc = Popen(qsub_cmd, shell=True, stdout=PIPE, stderr=PIPE)
+
+        # Adding proc.communicate call to wait for qsub to return and
+        # retrieve Torque ID from stdout.
+        # Communicate() pulls all stdout/stderr from the PIPEs
+        # This call waits until cmd is done
+        stdout, stderr = proc.communicate()
 
-    # Adding proc.communicate call to wait for qsub to return and
-    # retrieve Torque ID from stdout.
-    # Communicate() pulls all stdout/stderr from the PIPEs
-    # This call waits until cmd is done
-    stdout, stderr = proc.communicate()
+        # proc.returncode in this case means qsub successfully pushed the job
+        # onto Torque's queue.
+        #
+        # proc.returncode should always exist, but it will be equal to None if
+        # the proc hasn't finished yet. A negative value means the command was
+        # terminated by SIGNAL = the value (UNIX only)
+        if proc.returncode and proc.returncode != 0:
```

Comment:
Mainly it was just being conservative; this chunk of code was originally lifted from qiita-plugin-launcher-qsub. It wasn't hurting anything, but based on my experience, it will never short-circuit; qsub will always return immediately, and a process that's ended should always have a return code. If we're going to change it, we might as well just make it proc.returncode != 0: